tests/e2e/tests/e2e_util/util.py

```diff
@@ -21,102 +27,99 @@ def run(command: list, **kwargs) -> subprocess.CompletedProcess:
+        attempt = retry_state.attempt_number
+        LOG.info(
+            f"Attempt {attempt}/{retries} failed. Error: {retry_state.outcome.exception()}"
+        )
+        LOG.info(f"Retrying in {delay_s} seconds...")
+
+    _exceptions = exceptions or (Exception,)  # default to retry on all exceptions
+
+    class Retriable:
+        def __init__(self) -> None:
+            self.condition = None
+
+        @retry(
+            wait=wait_fixed(delay_s),
+            stop=stop_after_attempt(retries),
+            retry=retry_if_exception_type(_exceptions),
+            before_sleep=_before_sleep,
+        )
+        def exec(
+            self,
+            command_args: List[str],
+            harness: Optional[harness.Harness] = None,
+            instance: str = "",
+            **command_kwds,
+        ):
+            """
+            Execute a command against a harness or locally with subprocess to be retried.
+
+            :param  List[str]        command_args: The command to be executed, as a str or list of str
+            :param Map[str,str]      command_kwds: Additional keyword arguments to be passed to exec
+            :param Optional[Harness]      harness: test Harness object, to run the command on
+            :param str                   instance: Instance id in the test harness.
+            """
+
+            try:
+                if harness is not None:
+                    resp = harness.exec(
+                        instance, command_args, capture_output=True, **command_kwds
+                    )
                 else:
-                    # If all attempts fail, raise the last error
-                    raise e
+                    resp = subprocess.run(command_args, **command_kwds)
+            except subprocess.CalledProcessError as e:
+                LOG.error(f"  rc={e.returncode}")
+                LOG.error(f"  stdout={e.stdout.decode()}")
+                LOG.error(f"  stderr={e.stderr.decode()}")
+                raise
+            if self.condition:
+                assert self.condition(resp), "Failed to meet condition"
+
+        def until(
+            self, condition: Callable[[subprocess.CompletedProcess], bool] = None
+        ):
+            """
+            Test the output of the executed command against an expected response
+
+            :param Callable condition: a callable which returns a truth about the command output
+            """
+            self.condition = condition
+            return self
+
+    return Retriable()
 
 
 def setup_network(h: harness.Harness, instance_id: str):
     h.exec(instance_id, ["/snap/k8s/current/k8s/network-requirements.sh"])
 
     LOG.info("Waiting for network to be enabled...")
-    retry_until_condition(
-        h,
-        instance_id,
-        ["k8s", "enable", "network"],
-        condition=lambda p: "enabled" in p.stdout.decode(),
-    )
+    stubbornly(retries=15, delay_s=5).until(
+        lambda p: "enabled" in p.stdout.decode()
```

Comment:
should this be `lambda p: p.rc == 0`? or simply retry until success and no further check?

Reply:
Sure, that's kind of a fruitless check.  the rc should be enough